### PR TITLE
Discrete diffing

### PR DESCRIPTION
**Pull request by tbremer**
Took a pass at addressing issue #1 

This is very much open for comments and updates.

Additionally, I added:
- `.gitignore` file so that `node_modules` aren't published
- A script to automate counting lines (assumed comments and whitespace did not matter).
- A script to minify the code.

### CODE REVIEW

**Comment by tbremer** [comment]
Here is a file I used for testing my changes:

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <script defer src="/Teroy.js"></script>
</head>
<body>
    <div id="app"></div>
    <script>
        document.addEventListener('DOMContentLoaded', main);
        function main() {
            const myComponent = new Teroy("#app", {
                render() {
                    return `<h1>${this.data.title}</h1>
                    <input id="input" ${this.data.disabled}>
                    <button id="button" ${this.data.disabled}>Add to list</button>
                    
                    <ul>
                        ${this.data.list.map((listItem) => `
                            <li
                                id="${listItem.id}"
                                ${listItem.complete ? "data-is-complete" : ""}
                                class="item ${listItem.complete ? 'item-complete' : ''}"
                            >
                                ${listItem.value}
                                ${listItem.complete ? "(complete)" : ""}
                            </li>`).join("")}
                    </ul>`;
                },

                data: {
                    title: "To-do List",
                    disabled: "",
                    list: [],
                },
            });

            myComponent.show();

            myComponent.select("#button").addEventListener("click", () => {
                const inputValue = myComponent.select("#input").value;
                myComponent.data.list.push({value: inputValue, complete: false, id: (Math.round(Math.random() * Date.now())).toString(16)});

                if (myComponent.data.list.length >= 10) {
                    myComponent.data.title = "Done List!";
                    myComponent.data.disabled = "disabled";
                }
            });

            document.addEventListener('click', e => {
                if (!e.target.matches('.item')) return;
                const item = myComponent.data.list.find(i => i.id === e.target.id)

                item.complete = !item.complete;
                
                myComponent.update();
            })
        }
    </script>
</body>
</html>
```

**Comment by MathiasWP** [comment]
This looks very good Tom! I like your usage of the Set object, and it seems like this solves the problem as far as i can see. Some things i want to discuss with you:

Why do you fire the myComponent.update() function in the click eventListener in the HTML-test? Is this necessary? 

Should whitespace count when counting the lines? I kind of think it should, but also i understand why it doesn't really matter as whitespace is just used for readability. Should this be a part of the challenge to have the library be 100 lines of good, readable code? As the code is now, i think that's very possible to achieve. 

And what do you think about creating some test-scripts?

File: Teroy.js [hunk 1]
```diff
@@ -73,38 +73,67 @@
       this.rendered = true;
     }
 
+    /**
+     * 
+     * @param {NamedNodeMap} newAttrs 
+     * @param {NamedNodeMap} oldAttrs 
+     * @param {Node} node
+     */
+    diffAttributes(newAttrs, oldAttrs, node) {
```
Comment (MathiasWP) [inline Teroy.js:59]: In the finished code these comments should be removed. However, maybe we could create a Teroy.max.js file? Here we could in more depth comment and explain the code, and maybe we could create a flowchart to explain how/why everything works so automatically with the Proxy object?

**Review by MathiasWP — COMMENTED** [review]
And what's you thought about this? Considering how well this actually seems to work (and if we can do a lot of tests to find out the full potential of Teroy), i think that creating a website for this project could be something to look into.

**Comment by MathiasWP** [comment]
> Here is a file I used for testing my changes:
> 
> ```
> <!DOCTYPE html>
> <html lang="en">
> <head>
>     <meta charset="UTF-8">
>     <meta name="viewport" content="width=device-width, initial-scale=1.0">
>     <title>Document</title>
>     <script defer src="/Teroy.js"></script>
> </head>
> <body>
>     <div id="app"></div>
>     <script>
>         document.addEventListener('DOMContentLoaded', main);
>         function main() {
>             const myComponent = new Teroy("#app", {
>                 render() {
>                     return `<h1>${this.data.title}</h1>
>                     <input id="input" ${this.data.disabled}>
>                     <button id="button" ${this.data.disabled}>Add to list</button>
>                     
>                     <ul>
>                         ${this.data.list.map((listItem) => `
>                             <li
>                                 id="${listItem.id}"
>                                 ${listItem.complete ? "data-is-complete" : ""}
>                                 class="item ${listItem.complete ? 'item-complete' : ''}"
>                             >
>                                 ${listItem.value}
>                                 ${listItem.complete ? "(complete)" : ""}
>                             </li>`).join("")}
>                     </ul>`;
>                 },
> 
>                 data: {
>                     title: "To-do List",
>                     disabled: "",
>                     list: [],
>                 },
>             });
> 
>             myComponent.show();
> 
>             myComponent.select("#button").addEventListener("click", () => {
>                 const inputValue = myComponent.select("#input").value;
>                 myComponent.data.list.push({value: inputValue, complete: false, id: (Math.round(Math.random() * Date.now())).toString(16)});
> 
>                 if (myComponent.data.list.length >= 10) {
>                     myComponent.data.title = "Done List!";
>                     myComponent.data.disabled = "disabled";
>
>                     
>                     myComponent.data.list.length = 1
>       
>                 }
>             });
> 
>             document.addEventListener('click', e => {
>                 if (!e.target.matches('.item')) return;
>                 const item = myComponent.data.list.find(i => i.id === e.target.id)
>
>                 item.complete = !item.complete;
>             })
>         }
>     </script>
> </body>
> </html>
> ```

Here's an error we have to address. Changing the length of the data.list gives a "Cannot read property 'isEqualNode' of undefined" error. This problem does not occur with the current solution.

File: Teroy.js (hunk 1 above)
Comment (tbremer) [reply Teroy.js:59]: I like the idea of a `max` file, I can remove the comments and maybe we'll do another PR with the `max` version?

**Comment by tbremer** [comment]
> Why do you fire the myComponent.update() function in the click eventListener in the HTML-test? Is this necessary?

I think because of how the Proxy is setup. Nested values and methods on properties don't actually trigger the `set` handler.

e.g: `[].push` doesn't trigger set because the value isn't changing (IE: references same object). Same with the objects within that array. The proxy does not know those values are changing.

**Comment by tbremer** [comment]
>Should whitespace count when counting the lines? I kind of think it should, but also i understand why it doesn't really matter as whitespace is just used for readability. Should this be a part of the challenge to have the library be 100 lines of good, readable code? As the code is now, i think that's very possible to achieve.

Ok! I can update the line counter and do some more code-golfing to get the lines under 100.

**Comment by tbremer** [comment]
> Why do you fire the myComponent.update() function in the click eventListener in the HTML-test? Is this necessary?

I think because of how the Proxy is setup. Nested values and methods on properties don't actually trigger the `set` handler.

e.g: `[].push` doesn't trigger set because the value isn't changing (IE: references same object). Same with the objects within that array. The proxy does not know those values are changing.

>Should whitespace count when counting the lines? I kind of think it should, but also i understand why it doesn't really matter as whitespace is just used for readability. Should this be a part of the challenge to have the library be 100 lines of good, readable code? As the code is now, i think that's very possible to achieve.

Ok! I can update the line counter and do some more code-golfing to get the lines under 100.

>And what do you think about creating some test-scripts?

I don't mind setting up some tests. Do you have a preferred testing library?

**Comment by tbremer** [comment]
> Here's an error we have to address. Changing the length of the data.list gives a "Cannot read property 'isEqualNode' of undefined" error. This problem does not occur with the current solution.

I'll try to figure out where things went wrong here!

**Comment by MathiasWP** [comment]
The Proxy does actually know if a push or other functions have been run. This is because of the setup with the `get` handler. The `get` and `set` handler combination makes it so that Teroy runs everytime the data gets altered in any way. The `proxyPaused`value, combined with the `requestAnimationFrame` callback makes it so that it is possible to listen to all types of changes while preventing an infinite `get`-`set` loop. 

I am at least very sure that is the case, as i designed it to work this way - because i want the API to be as simple as possible. Try it out, and if it doesn’t work then issue it. 
You should be able to push, change the length or even change a specific item in an array (`array[2] = 'foo'`), and it'll be updated in the `this.parse(this.html.apply(this))` method.

**Comment by MathiasWP** [comment]
> > And what do you think about creating some test-scripts?
> 
> I don't mind setting up some tests. Do you have a preferred testing library?

I don't have any preferred testing library, so feel free to come with suggestions. After we've managed to find a solution to the discrete diffing can we begin working on the next things on our "todo list":

### Teroy To-Do list

- [ ] Create a discrete diffing algorithm so single items get updated in the DOM

- [ ] Set up test scripts; create tests

- [ ] Create a descriptive Teroy.max.js file with explanations. Flowchart?

- [ ] Set up a nice looking website for the library/framework

**Comment by MathiasWP** [comment]
I've created a flow-chart to explain how the Proxy handler works. GET and SET combined will notice any modification or use on an object, and here is the way Teroy handles this:
[code2flow_buAwdZ.pdf](https://github.com/MathiasWP/TeroyJS/files/4568684/code2flow_buAwdZ.pdf)

**Comment by tbremer** [comment]
Thanks, I didn’t realize a push would trigger a `get` within the proxy.

**Comment by MathiasWP** [comment]
Have you fixed the "Cannot read property 'isEqualNode' of undefined" error?

**Comment by MathiasWP** [comment]
I've added some more checks to the test file which can be used for making sure everything works as it does on the current version:

```html
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <script defer src="/Teroy.js"></script>
</head>

<body>
    <div id="app"></div>
    <script>
        document.addEventListener('DOMContentLoaded', main);
        function main() {
            const myComponent = new Teroy("#app", {
                render() {
                    return `<h1>${this.data.title}</h1>
                    <form id="input-form">
                        <input id="input" ${this.data.disabled}>
                        <button id="button" ${this.data.disabled}>Add to list</button>
                    </form>
                    <ul>
                        ${this.data.list.map((listItem) => `
                            <li
                                id="${listItem.id}"
                                ${listItem.complete ? "data-is-complete" : ""}
                                class="item ${listItem.complete ? 'item-complete' : ''}"
                            >
                                ${listItem.value}
                                ${listItem.complete ? "(complete)" : ""}
                            </li>`).join("")}
                    </ul>`;
                },

                data: {
                    title: "To-do List",
                    disabled: "",
                    list: [],
                    counter: 0
                },
            });

            myComponent.show();

            myComponent.select("#input-form").addEventListener("submit", (e) => {
                e.preventDefault()

                myComponent.data.counter += 1;

                const inputValue = myComponent.select("#input").value;
                myComponent.data.list.push({ value: inputValue, complete: false, id: (Math.round(Math.random() * Date.now())).toString(16) });

                if (myComponent.data.list.length === myComponent.data.counter && myComponent.data.counter >= 6) {
                    myComponent.data.title = "Done List!";
                    myComponent.data.disabled = "disabled";
                    myComponent.data.list[1] = ({ value: 'I WAS CHANGED', complete: false, id: (Math.round(Math.random() * Date.now())).toString(16) })

                    Object.assign(myComponent.data.list[0], { value: 'ASSIGNED' })

                    myComponent.data.list.length = 3;

                    const lastItem = myComponent.data.list.pop();
                    lastItem.value = 'I AM THE LAST ONE'
                    myComponent.data.list.push(lastItem)
                }
            });

            document.addEventListener('click', e => {
                if (!e.target.matches('.item')) return;
                const item = myComponent.data.list.find(i => i.id === e.target.id)

                item.complete = !item.complete;
            })
        }
    </script>
</body>

</html>
```

**Comment by tbremer** [comment]
Thanks for the test file, update to resolve the `isEqualNode` error.

**Comment by MathiasWP** [comment]
This looks good Tom! I'll merge it now.

**Comment by MathiasWP** [comment]
Oh btw, have you also minifed the new version?

**Pull request merged**